### ISSUE TITLE
Potential bug with LabelModel prediction

### ISSUE THREAD

**Issue by cdeepakroy** [snorkel-team/snorkel#1487]
## Issue description

I am trying to use snorkel to derive training labels for a binary text classification problem from 15 weakly supervised labeling functions (LF).

I am noticing lots of cases where 9/15 lfs abstained and remaining 6 LFs gave a label 0 but snorkel's final label probability was quite high and the binarized label was 1 (see table below)

![image](https://user-images.githubusercontent.com/1923534/66526801-b4cc2e80-eaae-11e9-97a3-1d881bb6f029.png)

Below are the learned weights from the label model for each LF

![image](https://user-images.githubusercontent.com/1923534/66527066-9c104880-eaaf-11e9-96b5-215788c32710.png)

Since all the 6 non-abstaining LFs gave a label 0, I expected the final label to be 0 as well and the probability values to be near 0.

Any clue what might be causing this behavior?

## Code example/repro steps

Since this is real data, it will be hard to reproduce this with a small code snippet. I can share anonymized label matrix.

## System info

* How you installed Snorkel (conda, pip, source): conda
* OS: Windows 10
* Python version: 3.7.4
* Snorkel version: 0.9.1

**Comment by cdeepakroy** [comment]
For the sake of debugging below are the values of `self.mu` and `self.p` after fitting the label model that are being used in the `LabelModel.predict_proba()` [here](https://github.com/snorkel-team/snorkel/blob/master/snorkel/labeling/model/label_model.py#L408-L410)

```
mu:
Parameter containing:
tensor([[3.0276e-01, 1.0000e-07],
        [4.7679e-02, 1.0000e-07],
        [1.9865e-01, 1.0000e-07],
        [7.5146e-03, 1.0000e-07],
        [1.6418e-01, 1.0000e-07],
        [1.0233e-03, 1.0000e-07],
        [7.3624e-01, 1.0000e-07],
        [3.6497e-03, 4.0087e-03],
        [7.1567e-01, 1.0000e-07],
        [1.6214e-03, 1.6699e-03],
        [6.5997e-01, 1.0000e-07],
        [5.6803e-04, 5.9173e-04],
        [7.3624e-01, 1.0000e-07],
        [3.6497e-03, 4.0087e-03],
        [7.1567e-01, 1.0000e-07],
        [1.6214e-03, 1.6699e-03],
        [6.5997e-01, 1.0000e-07],
        [5.6803e-04, 5.9173e-04],
        [4.1509e-01, 1.0000e+00],
        [6.8395e-04, 1.7432e-03],
        [4.0547e-01, 1.0000e+00],
        [2.1194e-04, 5.4538e-04],
        [3.8092e-01, 1.0000e+00],
        [4.2050e-05, 1.0374e-04],
        [4.1509e-01, 1.0000e+00],
        [6.8395e-04, 1.7432e-03],
        [4.0547e-01, 1.0000e+00],
        [2.1194e-04, 5.4538e-04],
        [3.8092e-01, 1.0000e+00],
        [4.2050e-05, 1.0374e-04]], requires_grad=True)
p:
[0.97259352 0.02740648]
```

**Comment by dataframing** [comment]
Can you share a redacted version of the summary table after calling `LFAnalysis` on your labeled development dataset?

**Comment by cdeepakroy** [comment]
Below are all the values of all the variables in the `predict_proba` function for a sample L matrix resembling the aforementioned issue

```
L:
[[-1 -1 -1 -1 -1 -1 -1 -1 -1  0  0  0  0  0  0]
 [-1 -1 -1 -1 -1 -1 -1 -1 -1  0  0  0  0  0  0]]

L_aug:
[[0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 1. 0. 1. 0.
  1. 0. 1. 0. 1. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 1. 0. 1. 0.
  1. 0. 1. 0. 1. 0.]]

jtm:
[[-1 -1 -1 -1 -1 -1 -1 -1 -1  0  0  0  0  0  0]
 [-1 -1 -1 -1 -1 -1 -1 -1 -1  0  0  0  0  0  0]]

np.log(mu):
[[-1.1948081e+00 -1.6118095e+01]
 [-3.0432580e+00 -1.6118095e+01]
 [-1.6162262e+00 -1.6118095e+01]
 [-4.8909011e+00 -1.6118095e+01]
 [-1.8067707e+00 -1.6118095e+01]
 [-6.8847365e+00 -1.6118095e+01]
 [-3.0620021e-01 -1.6118095e+01]
 [-5.6131086e+00 -5.5192933e+00]
 [-3.3453062e-01 -1.6118095e+01]
 [-6.4244943e+00 -6.3949699e+00]
 [-4.1555533e-01 -1.6118095e+01]
 [-7.4733391e+00 -7.4324613e+00]
 [-3.0620039e-01 -1.6118095e+01]
 [-5.6131086e+00 -5.5192933e+00]
 [-3.3453062e-01 -1.6118095e+01]
 [-6.4244943e+00 -6.3949699e+00]
 [-4.1555533e-01 -1.6118095e+01]
 [-7.4733391e+00 -7.4324613e+00]
 [-8.7926942e-01 -1.1920930e-07]
 [-7.2876229e+00 -6.3520508e+00]
 [-9.0269798e-01 -1.1920930e-07]
 [-8.4592047e+00 -7.5140243e+00]
 [-9.6516806e-01 -1.1920930e-07]
 [-1.0076643e+01 -9.1736116e+00]
 [-8.7926942e-01 -1.1920930e-07]
 [-7.2876229e+00 -6.3520508e+00]
 [-9.0269798e-01 -1.1920930e-07]
 [-8.4592047e+00 -7.5140243e+00]
 [-9.6516806e-01 -1.1920930e-07]
 [-1.0076643e+01 -9.1736116e+00]]

np.log(p):
[-0.02778904 -3.59697577]

X:
[[0.0039976  0.02740646]
 [0.0039976  0.02740646]]

Z:
[[0.03140407 0.03140407]
 [0.03140407 0.03140407]]

Y= X / Z:
[[0.12729576 0.87270424]
 [0.12729576 0.87270424]]

```

And below is the LF analysis output:

Polarity | Coverage | Overlaps | Conflicts | Correct | Incorrect | Emp. Acc. | Learned Weight
-- | -- | -- | -- | -- | -- | -- | --
  |   |   |   |   |   |   |  
[0, 1] | 0.315557 | 0.305014 | 0.046244 | 676207 | 194228 | 0.776861 | 0.933156
[0, 1] | 0.170503 | 0.170503 | 0.007580 | 432872 | 37444 | 0.920385 | 1.000000
[0, 1] | 0.131510 | 0.131510 | 0.001063 | 355573 | 7186 | 0.980191 | 1.000000
[0, 1] | 0.676211 | 0.676211 | 0.045135 | 1791017 | 74247 | 0.960195 | 1.000000
[0, 1] | 0.614154 | 0.614154 | 0.035245 | 1662048 | 32037 | 0.981089 | 1.000000
[0, 1] | 0.545063 | 0.545063 | 0.026004 | 1492398 | 11107 | 0.992613 | 1.000000
[0, 1] | 0.676211 | 0.676211 | 0.045135 | 1791017 | 74247 | 0.960195 | 1.000000
[0, 1] | 0.614154 | 0.614154 | 0.035245 | 1662048 | 32037 | 0.981089 | 1.000000
[0, 1] | 0.545063 | 0.545063 | 0.026004 | 1492398 | 11107 | 0.992613 | 1.000000
[0, 1] | 0.461756 | 0.461756 | 0.010750 | 464969 | 808741 | 0.365051 | 0.874396
[0, 1] | 0.439952 | 0.439952 | 0.008919 | 415178 | 798388 | 0.342114 | 0.896408
[0, 1] | 0.412456 | 0.412456 | 0.007203 | 340714 | 797006 | 0.299471 | 0.898235
[0, 1] | 0.461756 | 0.461756 | 0.010750 | 464969 | 808741 | 0.365051 | 0.874396
[0, 1] | 0.439952 | 0.439952 | 0.008919 | 415178 | 798388 | 0.342114 | 0.896408
[0, 1] | 0.412456 | 0.412456 | 0.007203 | 340714 | 797006 | 0.299471 | 0.89823

@dataframing Note that I dont have a dev set so I pass the predicted labels on training set to `lf_summary()` (see code snippet below)

```python
cb_est = [0.97259352 0.02740648]
lm = snorkel.labeling.LabelModel()
lm.fit(L_train, class_balance=cb_est, lr=0.01, n_epochs=2000, log_freq=100, seed=0)
y_pred = lm.predict(L_train)
df_lf_summary = snorkel.labeling.LFAnalysis(L_train).lf_summary(Y = y_pred, est_weights=lm.get_weights())
```

**Comment by dataframing** [comment]
@cdeepakroy A few observations:

1. From the original screenshot, it looks like your latter 6 labeling functions are all returning 0, whereas your first 9 are abstaining (-1).
2. The latter 6 labeling functions all exhibit significantly lower empirical accuracy/precision (~0.33?) compared to the initial 9 (~0.90?).
3. The learned weights are almost perfectly bi-modal and centered around either `0.87` or `1.0`.

I suppose I would also expect the model to return a more strongly negative label. Have you tried using an alternative labeler, like `snorkel.labeling.MajorityLabelVoter`? It should almost certainly give you negative labels for those examples you pointed out.

Otherwise, I'd look into doing a deeper dive using `pdb` or some other debugger to probe and understand exactly what components within `predict_proba` are contributing to the more strongly positive label. I tend to hack a PDB middleware models like so:

```python
import pdb

from snorkel.labeling import LabelModel
import numpy as np


class PDBLabelModel(LabelModel):

    def predict_proba(self, L: np.ndarray) -> np.ndarray:
        """Start PDB session, then step into superclass call."""
        pdb.set_trace()
        return super(LabelModel, self).predict_proba(L=L)
```

then proceed with your code as usual:

```python
# Initialize and fit the core label model.
lm = PDBLabelModel()
lm.fit(...)

# Get ready for the debugger to fire.
y_pred_proba = lm.predict_proba(L_train) 
```

Hope this helps!

**Comment by ajratner** [comment]
Hi @cdeepakroy thanks for trying out Snorkel!  From inspecting the learned labeling function conditional probabilities (`mu`), your `LabelModel` is learning a strong correlation between your last six LFs outputting `0` and the true label actually being `1`- in other words, it's learning that these labeling functions are most often *wrong* when outputting `0`.

*Why* is the `LabelModel` learning such a weird model?  It could be that the model was improperly trained or tuned.  It could be that your labeling functions are strongly (conditionally) correlated in weird ways (a problem that we've tackled on the research side, most recently in an ICML'19 paper, and built code for in other repos).  It could be some other complexity of your data or LFs, such as heavily data-conditional accuracies.  We've handled many of these more complex settings before, using other variants of the `LabelModel`, but unfortunately it's very hard to say what's happening without knowing more about your problem setup, data, LFs, and hyperparameter tuning procedure; and unfortunately this kind of diagnosis is outside the scope of what we can handle in the Issues page.  Happy to chat separately though!

Either way: In part prompted by this issue, we're currently discussing whether to prevent the model from ever learning models where the LFs are estimated to be wrong more often than right, as happened here.  Baking assumptions like these into the `LabelModel` can reduce its capability to handle interesting settings, but would also prevent results such as these, and put it more "on rails" by default... stay tuned!

**Comment by cdeepakroy** [comment]
Thanks for responding, @dataframing @ajratner 

@dataframing I narrowed down to the same observations you mentioned. For the erroneous cases that I mentioned above, indeed the majority vote label model would perform much better.

@ajratner To your point that the LabelModel may have been trained or tuned improperly, I am using lr=0.01 with a `constant` lr_schedule and n_epochs=2000. Below is the log of the loss after each epoch:

```
INFO:root:Computing O...
INFO:root:Estimating \mu...
INFO:root:[0 epochs]: TRAIN:[loss=15.680]
INFO:root:[100 epochs]: TRAIN:[loss=4.110]
INFO:root:[200 epochs]: TRAIN:[loss=3.829]
INFO:root:[300 epochs]: TRAIN:[loss=0.642]
INFO:root:[400 epochs]: TRAIN:[loss=0.085]
INFO:root:[500 epochs]: TRAIN:[loss=0.084]
INFO:root:[600 epochs]: TRAIN:[loss=0.084]
INFO:root:[700 epochs]: TRAIN:[loss=0.084]
INFO:root:[800 epochs]: TRAIN:[loss=0.084]
INFO:root:[900 epochs]: TRAIN:[loss=0.084]
INFO:root:[1000 epochs]: TRAIN:[loss=0.084]
INFO:root:[1100 epochs]: TRAIN:[loss=0.084]
INFO:root:[1200 epochs]: TRAIN:[loss=0.084]
INFO:root:[1300 epochs]: TRAIN:[loss=0.084]
INFO:root:[1400 epochs]: TRAIN:[loss=0.084]
INFO:root:[1500 epochs]: TRAIN:[loss=0.084]
INFO:root:[1600 epochs]: TRAIN:[loss=0.084]
INFO:root:[1700 epochs]: TRAIN:[loss=0.084]
INFO:root:[1800 epochs]: TRAIN:[loss=0.084]
INFO:root:[1900 epochs]: TRAIN:[loss=0.084]
INFO:root:Finished Training
```

The lowest loss after convergence is 0.084. Based on your experiments, Is the loss expected to be much lower than this or is this typical? I briefly experimented with other lr shceulers but could not get the loss to drop below this.

Could you point me these other repos (if they are open to public) that handle these complex settings.

**Comment by timellemeet** [comment]
@cdeepakroy I'm experiencing similar behavior. Have you gained any new insights for your use case?
